Mathlib/LinearAlgebra/FreeModule/PID.lean

```diff
@@ -440,6 +441,65 @@ structure Basis.SmithNormalForm (N : Submodule R M) (ι : Type*) (n : ℕ) where
   snf : ∀ i, (bN i : M) = a i • bM (f i)
 #align basis.smith_normal_form Basis.SmithNormalForm
 
+namespace Basis.SmithNormalForm
+
+variable {n : ℕ} {N : Submodule R M} (snf : Basis.SmithNormalForm N ι n) (m : N)
+
+lemma repr_eq_zero_of_nmem_range {i : ι} (hi : i ∉ Set.range snf.f) :
+    snf.bM.repr m i = 0 := by
+  obtain ⟨m, hm⟩ := m
+  obtain ⟨c, rfl⟩ := snf.bN.mem_submodule_iff.mp hm
+  replace hi : ∀ j, snf.f j ≠ i := by simpa using hi
+  simp [Finsupp.single_apply, hi, snf.snf]
+
+@[simp] lemma repr_comp_embedding_eq_smul :
+    snf.bM.repr m ∘ snf.f = (snf.bN.repr m : Fin n → R) • snf.a := by
+  ext j
+  obtain ⟨m, hm⟩ := m
+  obtain ⟨c, rfl⟩ := snf.bN.mem_submodule_iff.mp hm
+  replace hm : (⟨Finsupp.sum c fun i t ↦ t • (↑(snf.bN i) : M), hm⟩ : N) =
+      Finsupp.sum c fun i t ↦ t • ⟨snf.bN i, (snf.bN i).2⟩ := by ext; change _ = N.subtype _; simp
+  simp_rw [hm, LinearEquiv.map_finsupp_sum, map_smul, Subtype.coe_eta, repr_self,
+    Finsupp.smul_single, smul_eq_mul, mul_one, Finsupp.sum_single, Pi.mul_apply]
+  classical
+  simp_rw [snf.snf, map_smul, repr_self, Finsupp.smul_single, smul_eq_mul, mul_one,
+    Function.comp_apply, Finsupp.sum_apply, Finsupp.single_apply, EmbeddingLike.apply_eq_iff_eq,
+    Finsupp.sum_ite_eq', Finsupp.mem_support_iff, ite_not, ite_eq_right_iff]
+  exact fun a ↦ (mul_eq_zero_of_left a _).symm
+
+@[simp] lemma repr_apply_embedding_eq_repr_smul {i : Fin n} :
+    snf.bM.repr m (snf.f i) = snf.bN.repr (snf.a i • m) i := by
+  change (snf.bM.repr m ∘ snf.f) i = _
+  rw [repr_comp_embedding_eq_smul, Pi.smul_apply', smul_eq_mul, mul_comm _ (snf.a i), map_smul,
+    Finsupp.coe_smul, Pi.smul_apply, smul_eq_mul]
+
+lemma le_ker_coord_of_nmem_range {i : ι} (hi : i ∉ Set.range snf.f) :
+    N ≤ LinearMap.ker (snf.bM.coord i) :=
+  fun m hm ↦ snf.repr_eq_zero_of_nmem_range ⟨m, hm⟩ hi
+
+@[simp] lemma coord_apply_embedding_eq_smul_coord {i : Fin n} :
+    snf.bM.coord (snf.f i) ∘ₗ N.subtype = snf.a i • snf.bN.coord i := by
+  ext m
+  rw [LinearMap.smul_apply, LinearMap.coe_comp, Submodule.coeSubtype, Function.comp_apply,
+    coord_apply, coord_apply, snf.repr_apply_embedding_eq_repr_smul m, map_smul, Finsupp.coe_smul,
+    Pi.smul_apply]
```

Comment:
It looks like you use only `simp` lemmas here. Can this just be:
```suggestion
  simp
```

Reply:
Sadly no; I too would like if this could be just `simp`.

The first issue is minor: we need to make sure `smul_eq_mul` and `Algebra.id.smul_eq_mul` don't fire (or else we need to rethink our simp set elsewhere). However `simp [-smul_eq_mul, -Algebra.id.smul_eq_mul]` still fails because `simp` won't fire `Pi.smul_apply`. I don't know why!

For emphasis: even replacing the `rw` with `simp only`, everything fires except `Pi.smul_apply`.

Reply:
Got it to work at the cost of having to feed in `Pi.smul_apply _` to `simp`.